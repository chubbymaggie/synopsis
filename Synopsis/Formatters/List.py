#
# Copyright (C) 2006 Stefan Seefeld
# All rights reserved.
# Licensed to the public under the terms of the GNU LGPL (>= 2),
# see the file COPYING for details.
#

from Synopsis.Processor import Processor, Parameter
from Synopsis import AST

class Formatter(Processor, AST.Visitor):
    """Generate a high-level list of the content of a syn file.
    This formatter can lists source files (by name), as well as
    declarations (by name, type) contained in a given scope."""

    show_files = Parameter(False, 'list files')
    show_scope = Parameter(None, 'list declarations in the given scope')

    scope_separator = {'C' : '::',
                       'C++' : '::',
                       'IDL' : '::',
                       'Python' : '.'}

    def process(self, ast, **kwds):

        self.set_parameters(kwds)
        self.ast = self.merge_input(ast)

        if self.show_files:
            for name, sf in self.ast.files().iteritems():
                print name, sf.name

        if self.show_scope is not None:
            if '.' in self.show_scope:
                self.scope = tuple(self.show_scope.split('.'))
            else:
                self.scope = tuple(self.show_scope.split('::'))
                
            # Now guess the current language by looking at the
            # first declaration.
            d = len(self.ast.declarations()) and self.ast.declarations()[0]
            sf = d and d.file() or None
            if not sf: # d was a MetaModule, so let's assume C++ or IDL.
                self.sep = '::'
            else:
                lang = sf.annotations.get('language', 'Python')
                self.sep = self.scope_separator[lang]


            for d in self.ast.declarations():
                d.accept(self)
                
        return self.ast


    def visitScope(self, node):

        if self.scope == node.name():

            # We found the right scope.
            # List all declarations directly contained in it.
            declarations = node.declarations()[:]
            declarations.sort(lambda x, y : cmp(x.name(), y.name()))
            for d in declarations:
                if isinstance(d, AST.Builtin): continue
                name = d.name()[-1]
                _type = d.type()
                print '%s : %s'%(name, _type)
        elif (len(node.name()) < self.scope and
              self.scope[0:len(node.name())] == node.name()):

            # We found a parent scope.
            # Visit child scopes.
            for d in node.declarations():
                d.accept(self)


__all__ = ['Formatter']
