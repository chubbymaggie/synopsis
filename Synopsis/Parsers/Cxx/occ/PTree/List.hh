//
// Copyright (C) 2004 Stefan Seefeld
// All rights reserved.
// Licensed to the public under the terms of the GNU LGPL (>= 2),
// see the file COPYING for details.
//
#ifndef _PTree_List_hh
#define _PTree_List_hh

#include <PTree/Node.hh>

namespace PTree
{

class List : public Node
{
public:
  List(Node *p, Node *q) : Node(p, q) {}
  bool is_atom() const { return false;}
  virtual void accept(Visitor *visitor) { visitor->visit(this);}

  virtual void write(std::ostream &) const;
  virtual void print(std::ostream &, size_t, size_t) const;

  int Write(std::ostream&, int);  
protected:
  bool too_deep(std::ostream&, size_t) const;
  void print_encoded(std::ostream &, size_t, size_t) const;
};

}

#endif
