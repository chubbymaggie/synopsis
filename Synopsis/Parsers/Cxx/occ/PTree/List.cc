//
// Copyright (C) 2004 Stefan Seefeld
// All rights reserved.
// Licensed to the public under the terms of the GNU LGPL (>= 2),
// see the file COPYING for details.
//

#include <PTree/List.hh>
#include <Encoding.hh>
#include <cassert>
#include <stdexcept>

using namespace PTree;

void List::write(std::ostream &os) const
{
  assert(this);
  for (const Node *p = this; p; p = p->cdr())
  {
    const Node *data = p->car();
    if (data) data->write(os);
    else if(p->is_atom()) throw std::logic_error("List::write(): not list");
  }
}

void List::print(std::ostream &os, size_t indent, size_t depth) const
{
  if(too_deep(os, depth)) return;

  const Node *rest = this;
  os << '[';
  while(rest != 0)
  {
    if(rest->is_atom())
    {
      os << "@ ";
      rest->print(os, indent, depth + 1);
      rest = 0;
    }
    else
    {
      const Node *head = rest->car();
      if(head == 0) os << "nil";
      else head->print(os, indent, depth + 1);
      rest = rest->cdr();
      if(rest != 0) os << ' ';
    }
  }
  os << ']';
}

void List::print_encoded(std::ostream &os, size_t indent, size_t depth) const
{
  if (show_encoded)
  {
    const char *encode = encoded_type();
    if(encode)
    {
      os << '#';
      Encoding::print(os, encode);
    }
    encode = encoded_name();
    if(encode)
    {
      os << '@';
      Encoding::print(os, encode);
    }
  }
  List::print(os, indent, depth);
}

bool List::too_deep(std::ostream& s, size_t depth) const
{
  if(depth >= 32)
  {
    s << " ** too many nestings ** ";
    return true;
  }
  else return false;
}

int List::Write(std::ostream& out, int indent)
{
  int n = 0;
  Node *p = this;
  while(true)
  {
    Node *head = p->car();
    if(head != 0) n += head->Write(out, indent);

    p = p->cdr();
    if(p == 0) break;
    else if(p->is_atom())
      throw std::runtime_error("List::Write(): not list");
    else out << ' ';
  }
  return n;
}
