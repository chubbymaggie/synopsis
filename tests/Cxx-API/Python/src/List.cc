#include <Synopsis/List.hh>
#include "Guard.hh"
#include <string>
#include <iostream>

using namespace Synopsis;

void test1()
{
  List list;
  list.append('a');
  list.append(1);
  list.append("hello");
  list.append("world");
  std::cout << Object::narrow<std::string>(list.str()) << std::endl;
  list.sort();
  std::cout << Object::narrow<std::string>(list.str()) << std::endl;
  list.reverse();
  std::cout << Object::narrow<std::string>(list.str()) << std::endl;
  Object o = list.get(0);
  std::cout << Object::narrow<std::string>(o.str()) << std::endl;

  for (List::iterator i = list.begin(); i != list.end(); ++i)
    std::cout << Object::narrow<std::string>(i->str()) << std::endl;
}

int main(int, char **)
{  
  try
  {
    test1();
  }
  catch (const std::exception &e)
  {
    std::cout << "Error : " << e.what() << std::endl;
  }
}
