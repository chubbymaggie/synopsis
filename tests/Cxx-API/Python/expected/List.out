['a', 1, 'hello', 'world']
[1, 'a', 'hello', 'world']
['world', 'hello', 'a', 1]
world
world
hello
a
1
